File: sdks/python/agentrace/_emitter.py
from __future__ import annotations

import time
import uuid
from contextlib import contextmanager
from datetime import datetime, timezone
from typing import Any, Dict, Generator, List, Optional

from ._models import (
    AgentInfo,
    AgentTrace,
    AnyStep,
    CustomStep,
    ErrorStep,
    HandoffStep,
    LoopIteration,
    LoopStep,
    MemoryStep,
    RespondStep,
    ThinkStep,
    TokenCounts,
    ToolCallStep,
)


def _now() -> datetime:
    return datetime.now(timezone.utc)


def _step_id() -> str:
    return f"step_{uuid.uuid4().hex[:8]}"


def _trace_id() -> str:
    return f"trace_{uuid.uuid4().hex[:12]}"


class _PendingToolCall:
    """Holds the mutable state of a timed tool call before it is committed."""

    def __init__(self) -> None:
        self.output: Any = None
        self.error: Optional[str] = None


class Emitter:
    """Instrument an agent run and produce an AgentTrace.

    Usage::

        emitter = Emitter(agent="my-agent", goal="Find top competitors")
        emitter.think("I should search first")
        emitter.tool_call("web_search", {"query": "AI agents 2026"}, output={...}, duration_ms=1200)
        emitter.respond("Here are the results", format="markdown", final=True)
        trace = emitter.finish()
        trace.save("run.atrace")

    For auto-timed tool calls::

        with emitter.time_tool_call("web_fetch", {"url": "https://..."}) as call:
            result = fetch(...)
            call.output = result
        # duration_ms is recorded automatically
    """

    def __init__(
        self,
        *,
        agent: str,
        goal: str,
        model: Optional[str] = None,
        agent_version: Optional[str] = None,
        input: Optional[str] = None,
        parent_trace_id: Optional[str] = None,
        trace_id: Optional[str] = None,
        meta: Optional[Dict[str, Any]] = None,
    ) -> None:
        self._agent = AgentInfo(name=agent, model=model, version=agent_version)
        self._goal = goal
        self._input = input
        self._parent_trace_id = parent_trace_id
        self._trace_id = trace_id or _trace_id()
        self._meta = meta
        self._started_at = _now()
        self._steps: List[AnyStep] = []

    # ------------------------------------------------------------------
    # Step methods
    # ------------------------------------------------------------------

    def think(
        self,
        reasoning: str,
        *,
        meta: Optional[Dict[str, Any]] = None,
        step_id: Optional[str] = None,
    ) -> ThinkStep:
        step = ThinkStep(
            id=step_id or _step_id(),
            started_at=_now(),
            reasoning=reasoning,
            meta=meta,
        )
        self._steps.append(step)
        return step

    def tool_call(
        self,
        tool: str,
        input: Dict[str, Any],
        *,
        output: Any = None,
        duration_ms: Optional[float] = None,
        error: Optional[str] = None,
        retry_of: Optional[str] = None,
        meta: Optional[Dict[str, Any]] = None,
        step_id: Optional[str] = None,
    ) -> ToolCallStep:
        now = _now()
        step = ToolCallStep(
            id=step_id or _step_id(),
            started_at=now,
            ended_at=now if duration_ms is not None else None,
            tool=tool,
            input=input,
            output=output,
            duration_ms=duration_ms,
            error=error,
            retry_of=retry_of,
            meta=meta,
        )
        self._steps.append(step)
        return step

    @contextmanager
    def time_tool_call(
        self,
        tool: str,
        input: Dict[str, Any],
        *,
        retry_of: Optional[str] = None,
        meta: Optional[Dict[str, Any]] = None,
        step_id: Optional[str] = None,
    ) -> Generator[_PendingToolCall, None, None]:
        """Context manager that auto-records duration_ms and ended_at."""
        sid = step_id or _step_id()
        started = _now()
        t0 = time.perf_counter()
        pending = _PendingToolCall()
        try:
            yield pending
        except Exception as exc:
            if pending.error is None:
                pending.error = str(exc)
            raise
        finally:
            elapsed_ms = round((time.perf_counter() - t0) * 1000, 1)
            step = ToolCallStep(
                id=sid,
                started_at=started,
                ended_at=_now(),
                tool=tool,
                input=input,
                output=pending.output,
                duration_ms=elapsed_ms,
                error=pending.error,
                retry_of=retry_of,
                meta=meta,
            )
            self._steps.append(step)

    def respond(
        self,
        content: str,
        *,
        format: Optional[str] = None,
        final: bool = False,
        meta: Optional[Dict[str, Any]] = None,
        step_id: Optional[str] = None,
    ) -> RespondStep:
        step = RespondStep(
            id=step_id or _step_id(),
            started_at=_now(),
            content=content,
            format=format,
            final=final or None,
            meta=meta,
        )
        self._steps.append(step)
        return step

    def handoff(
        self,
        to_agent: str,
        task: str,
        *,
        child_trace_id: Optional[str] = None,
        meta: Optional[Dict[str, Any]] = None,
        step_id: Optional[str] = None,
    ) -> HandoffStep:
        step = HandoffStep(
            id=step_id or _step_id(),
            started_at=_now(),
            to_agent=to_agent,
            task=task,
            child_trace_id=child_trace_id,
            meta=meta,
        )
        self._steps.append(step)
        return step

    def error(
        self,
        message: str,
        *,
        code: Optional[str] = None,
        recoverable: Optional[bool] = None,
        meta: Optional[Dict[str, Any]] = None,
        step_id: Optional[str] = None,
    ) -> ErrorStep:
        step = ErrorStep(
            id=step_id or _step_id(),
            started_at=_now(),
            message=message,
            code=code,
            recoverable=recoverable,
            meta=meta,
        )
        self._steps.append(step)
        return step

    def memory(
        self,
        operation: str,
        *,
        store: Optional[str] = None,
        key: Optional[str] = None,
        value: Any = None,
        query: Optional[str] = None,
        results: Optional[List[Any]] = None,
        duration_ms: Optional[float] = None,
        error: Optional[str] = None,
        meta: Optional[Dict[str, Any]] = None,
        step_id: Optional[str] = None,
    ) -> MemoryStep:
        step = MemoryStep(
            id=step_id or _step_id(),
            started_at=_now(),
            operation=operation,
            store=store,
            key=key,
            value=value,
            query=query,
            results=results,
            duration_ms=duration_ms,
            error=error,
            meta=meta,
        )
        self._steps.append(step)
        return step

    def loop(
        self,
        iterations: List[List[AnyStep]],
        *,
        label: Optional[str] = None,
        max_iterations: Optional[int] = None,
        exit_reason: Optional[str] = None,
        meta: Optional[Dict[str, Any]] = None,
        step_id: Optional[str] = None,
    ) -> LoopStep:
        loop_iterations = [
            LoopIteration(index=i + 1, steps=steps)
            for i, steps in enumerate(iterations)
        ]
        step = LoopStep(
            id=step_id or _step_id(),
            started_at=_now(),
            label=label,
            max_iterations=max_iterations,
            exit_reason=exit_reason,
            iterations=loop_iterations,
            meta=meta,
        )
        self._steps.append(step)
        return step

    def custom(
        self,
        custom_type: str,
        *,
        data: Optional[Dict[str, Any]] = None,
        meta: Optional[Dict[str, Any]] = None,
        step_id: Optional[str] = None,
    ) -> CustomStep:
        step = CustomStep(
            id=step_id or _step_id(),
            started_at=_now(),
            custom_type=custom_type,
            data=data,
            meta=meta,
        )
        self._steps.append(step)
        return step

    # ------------------------------------------------------------------
    # Finalise
    # ------------------------------------------------------------------

    def finish(
        self,
        status: str = "completed",
        *,
        tokens: Optional[Dict[str, int]] = None,
    ) -> AgentTrace:
        """Close the trace and return an AgentTrace object."""
        token_counts = TokenCounts(**tokens) if tokens else None
        return AgentTrace(
            atrace="0.2.0",
            id=self._trace_id,
            agent=self._agent,
            goal=self._goal,
            input=self._input,
            status=status,
            started_at=self._started_at,
            ended_at=_now(),
            steps=self._steps,
            tokens=token_counts,
            parent_trace_id=self._parent_trace_id,
            meta=self._meta,
        )

    def fail(self, *, tokens: Optional[Dict[str, int]] = None) -> AgentTrace:
        """Shorthand for finish(status='failed')."""
        return self.finish("failed", tokens=tokens)

File: sdks/typescript/src/emitter.ts
import { randomBytes } from "crypto";
import {
  AgentInfo,
  AgentTrace,
  AnyStep,
  CustomStep,
  ErrorStep,
  HandoffStep,
  LoopIteration,
  LoopStep,
  MemoryOperation,
  MemoryStep,
  RespondStep,
  ResponseFormat,
  ThinkStep,
  TokenCounts,
  ToolCallStep,
} from "./types";

function now(): string {
  return new Date().toISOString();
}

function stepId(): string {
  return `step_${randomBytes(4).toString("hex")}`;
}

function traceId(): string {
  return `trace_${randomBytes(6).toString("hex")}`;
}

export interface EmitterOptions {
  agent: string;
  goal: string;
  model?: string;
  agentVersion?: string;
  input?: string;
  parentTraceId?: string;
  traceId?: string;
  meta?: Record<string, unknown>;
}

export interface PendingToolCall {
  output?: unknown;
  error?: string | null;
}

/**
 * Instrument an agent run and produce an AgentTrace.
 *
 * @example
 * ```ts
 * import { Emitter } from "agentrace";
 *
 * const emitter = new Emitter({ agent: "my-agent", goal: "Find competitors" });
 * emitter.think("I should search first");
 * emitter.toolCall("web_search", { query: "AI agents" }, { output: { results: [] }, durationMs: 1200 });
 * emitter.respond("Here are the results", { format: "markdown", final: true });
 * const trace = emitter.finish();
 * ```
 */
export class Emitter {
  private readonly _agent: AgentInfo;
  private readonly _goal: string;
  private readonly _input?: string;
  private readonly _parentTraceId?: string;
  private readonly _traceId: string;
  private readonly _meta?: Record<string, unknown>;
  private readonly _startedAt: string;
  private readonly _steps: AnyStep[] = [];

  constructor(opts: EmitterOptions) {
    this._agent = {
      name: opts.agent,
      model: opts.model,
      version: opts.agentVersion,
    };
    this._goal = opts.goal;
    this._input = opts.input;
    this._parentTraceId = opts.parentTraceId;
    this._traceId = opts.traceId ?? traceId();
    this._meta = opts.meta;
    this._startedAt = now();
  }

  think(
    reasoning: string,
    opts?: { meta?: Record<string, unknown>; stepId?: string }
  ): ThinkStep {
    const step: ThinkStep = {
      id: opts?.stepId ?? stepId(),
      type: "think",
      started_at: now(),
      reasoning,
      meta: opts?.meta,
    };
    this._steps.push(step);
    return step;
  }

  toolCall(
    tool: string,
    input: Record<string, unknown>,
    opts?: {
      output?: unknown;
      durationMs?: number;
      error?: string | null;
      retryOf?: string;
      meta?: Record<string, unknown>;
      stepId?: string;
    }
  ): ToolCallStep {
    const t = now();
    const step: ToolCallStep = {
      id: opts?.stepId ?? stepId(),
      type: "tool_call",
      started_at: t,
      ended_at: opts?.durationMs != null ? now() : undefined,
      tool,
      input,
      output: opts?.output,
      duration_ms: opts?.durationMs,
      error: opts?.error ?? undefined,
      retry_of: opts?.retryOf,
      meta: opts?.meta,
    };
    this._steps.push(step);
    return step;
  }

  /**
   * Auto-timed tool call using async/await.
   *
   * @example
   * ```ts
   * const result = await emitter.timeToolCall("web_fetch", { url }, async (call) => {
   *   const data = await fetch(url);
   *   call.output = data;
   * });
   * ```
   */
  async timeToolCall(
    tool: string,
    input: Record<string, unknown>,
    fn: (call: PendingToolCall) => Promise<void>,
    opts?: { retryOf?: string; meta?: Record<string, unknown>; stepId?: string }
  ): Promise<ToolCallStep> {
    const sid = opts?.stepId ?? stepId();
    const startedAt = now();
    const t0 = Date.now();
    const pending: PendingToolCall = {};
    let errorMsg: string | null = null;

    try {
      await fn(pending);
    } catch (err) {
      errorMsg = err instanceof Error ? err.message : String(err);
      throw err;
    } finally {
      const durationMs = Date.now() - t0;
      const step: ToolCallStep = {
        id: sid,
        type: "tool_call",
        started_at: startedAt,
        ended_at: now(),
        tool,
        input,
        output: pending.output,
        duration_ms: durationMs,
        error: errorMsg ?? pending.error ?? undefined,
        retry_of: opts?.retryOf,
        meta: opts?.meta,
      };
      this._steps.push(step);
      return step;
    }
  }

  respond(
    content: string,
    opts?: {
      format?: ResponseFormat;
      final?: boolean;
      meta?: Record<string, unknown>;
      stepId?: string;
    }
  ): RespondStep {
    const step: RespondStep = {
      id: opts?.stepId ?? stepId(),
      type: "respond",
      started_at: now(),
      content,
      format: opts?.format,
      final: opts?.final,
      meta: opts?.meta,
    };
    this._steps.push(step);
    return step;
  }

  handoff(
    toAgent: string,
    task: string,
    opts?: {
      childTraceId?: string;
      meta?: Record<string, unknown>;
      stepId?: string;
    }
  ): HandoffStep {
    const step: HandoffStep = {
      id: opts?.stepId ?? stepId(),
      type: "handoff",
      started_at: now(),
      to_agent: toAgent,
      task,
      child_trace_id: opts?.childTraceId,
      meta: opts?.meta,
    };
    this._steps.push(step);
    return step;
  }

  error(
    message: string,
    opts?: {
      code?: string;
      recoverable?: boolean;
      meta?: Record<string, unknown>;
      stepId?: string;
    }
  ): ErrorStep {
    const step: ErrorStep = {
      id: opts?.stepId ?? stepId(),
      type: "error",
      started_at: now(),
      message,
      code: opts?.code,
      recoverable: opts?.recoverable,
      meta: opts?.meta,
    };
    this._steps.push(step);
    return step;
  }

  memory(
    operation: MemoryOperation,
    opts?: {
      store?: string;
      key?: string;
      value?: unknown;
      query?: string;
      results?: unknown[];
      durationMs?: number;
      error?: string | null;
      meta?: Record<string, unknown>;
      stepId?: string;
    }
  ): MemoryStep {
    const step: MemoryStep = {
      id: opts?.stepId ?? stepId(),
      type: "memory",
      started_at: now(),
      operation,
      store: opts?.store,
      key: opts?.key,
      value: opts?.value,
      query: opts?.query,
      results: opts?.results,
      duration_ms: opts?.durationMs,
      error: opts?.error ?? undefined,
      meta: opts?.meta,
    };
    this._steps.push(step);
    return step;
  }

  loop(
    iterations: AnyStep[][],
    opts?: {
      label?: string;
      maxIterations?: number;
      exitReason?: string;
      meta?: Record<string, unknown>;
      stepId?: string;
    }
  ): LoopStep {
    const loopIterations: LoopIteration[] = iterations.map((steps, i) => ({
      index: i + 1,
      steps,
    }));
    const step: LoopStep = {
      id: opts?.stepId ?? stepId(),
      type: "loop",
      started_at: now(),
      label: opts?.label,
      max_iterations: opts?.maxIterations,
      exit_reason: opts?.exitReason,
      iterations: loopIterations,
      meta: opts?.meta,
    };
    this._steps.push(step);
    return step;
  }

  custom(
    customType: string,
    opts?: {
      data?: Record<string, unknown>;
      meta?: Record<string, unknown>;
      stepId?: string;
    }
  ): CustomStep {
    const step: CustomStep = {
      id: opts?.stepId ?? stepId(),
      type: "custom",
      started_at: now(),
      custom_type: customType,
      data: opts?.data,
      meta: opts?.meta,
    };
    this._steps.push(step);
    return step;
  }

  finish(
    status: "completed" | "failed" | "cancelled" = "completed",
    opts?: { tokens?: TokenCounts }
  ): AgentTrace {
    return {
      atrace: "0.2.0",
      id: this._traceId,
      agent: this._agent,
      goal: this._goal,
      input: this._input,
      status,
      started_at: this._startedAt,
      ended_at: now(),
      steps: [...this._steps],
      tokens: opts?.tokens,
      parent_trace_id: this._parentTraceId,
      meta: this._meta,
    };
  }

  fail(opts?: { tokens?: TokenCounts }): AgentTrace {
    return this.finish("failed", opts);
  }
}
